### PR TITLE
Allow for subdirectory structure within data directory

### PR DESCRIPTION
**Pull request by PfisterAn**
- recursively read the data directory
  - @TODO/SUGGEST - may be worth to think about using the sub directory information for enriching the index and/or use it for access control (e.g., subdirectory names may be named like the Azure AD Group for access control)

## Purpose

* Ability to bring subfolder structure into data directory for better arranging the data, at least on the directory level for now

## Does this introduce a breaking change?

```
[ ] Yes
[ x] No
```

## Pull Request Type
What kind of change does this Pull Request introduce?

```
[ ] Bugfix
[ x] Feature
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Documentation content changes
[ ] Other... Please describe:
```

## How to Test
*  Run prepdocs on the data directory where content is arranged within subdirectories

```
git clone [repo-address]
cd [repo-name]
git checkout [branch-name]
npm install
```

* Test the code

```
```

## What to Check
Verify that the following are valid
* ...

### CODE REVIEW

File: scripts/prepdocs.py [hunk 1]
```diff
@@ -314,6 +314,29 @@ def refresh_openai_token():
...
         open_ai_token_cache[CACHE_KEY_CREATED_TIME] = time.time()
 
+# recursively read the directory to allow for a subdirectory structure
+def read_files(location, use_vectors):
+    for filename in glob.glob(location):
+        if args.verbose: print(f"Processing '{filename}'")
+        if args.remove:
+            remove_blobs(filename)
+            remove_from_index(filename)
+        elif args.removeall:
```
Comment (pamelafox) [inline scripts/prepdocs.py:324]: This was already an issue, but should this if here be outside the function?

File: scripts/prepdocs.py [hunk 2]
```diff
@@ -314,6 +314,29 @@ def refresh_openai_token():
         openai.api_key = token_cred.get_token("https://cognitiveservices.azure.com/.default").token
         open_ai_token_cache[CACHE_KEY_CREATED_TIME] = time.time()
 
+# recursively read the directory to allow for a subdirectory structure
+def read_files(location, use_vectors):
```
Comment (pamelafox) [inline scripts/prepdocs.py:318]: Could you type this to make it clear whether location is a string or a path-like object?

**Comment by pamelafox** [comment]
Thanks, this seems useful!

File: scripts/prepdocs.py (hunk 1 above)
Comment (PfisterAn) [reply scripts/prepdocs.py:327]: Agree, thought the wrong way here, at least the removeall needs to be outside the function, which already is, so its redundant actually and/or never executed anyways, will be removed!

File: scripts/prepdocs.py (hunk 2 above)
Comment (PfisterAn) [reply scripts/prepdocs.py]: Should be fine now now, renamed from `location` to `pathPattern`

File: scripts/prepdocs.py [hunk 3]
```diff
@@ -314,6 +314,29 @@ def refresh_openai_token():
         openai.api_key = token_cred.get_token("https://cognitiveservices.azure.com/.default").token
         open_ai_token_cache[CACHE_KEY_CREATED_TIME] = time.time()
 
+# recursively read the directory to allow for a subdirectory structure
```
[the remark this answers is not in the document]
Comment (PfisterAn) [reply scripts/prepdocs.py]: Done!

File: scripts/prepdocs.py (hunk 1 above)
Comment (PfisterAn) [reply scripts/prepdocs.py]: This got fixed!

File: scripts/prepdocs.py [hunk 4]
```diff
@@ -314,6 +314,29 @@ def refresh_openai_token():
         openai.api_key = token_cred.get_token("https://cognitiveservices.azure.com/.default").token
         open_ai_token_cache[CACHE_KEY_CREATED_TIME] = time.time()
 
+"""Readfile function
+
+recursively read directory structure under `pathPattern` and execute indexing for the individual files
+"""
+def read_files(pathPattern, use_vectors):
```
Comment (pamelafox) [inline scripts/prepdocs.py:321]: pathPattern should be path_pattern, per naming conventions detailed at pep8.org

Please also add a type annotation- learn more here: https://realpython.com/lessons/annotations/

File: scripts/prepdocs.py [hunk 5]
```diff
@@ -314,6 +314,29 @@ def refresh_openai_token():
         openai.api_key = token_cred.get_token("https://cognitiveservices.azure.com/.default").token
         open_ai_token_cache[CACHE_KEY_CREATED_TIME] = time.time()
 
+"""Readfile function
```
Comment (pamelafox) [inline scripts/prepdocs.py:317]: Please move inside the function and remove top line, ie.

def read_files(path_pattern, use_vectors):
  """
  Recursively read directory structure under `path_pattern`
  and execute indexing for the individual files
  """

File: scripts/prepdocs.py [hunk 6]
```diff
@@ -379,17 +402,4 @@ def refresh_openai_token():
             create_search_index()
 
         print("Processing files...")
-        for filename in glob.glob(args.files):
-            if args.verbose: print(f"Processing '{filename}'")
-            if args.remove:
-                remove_blobs(filename)
-                remove_from_index(filename)
-            elif args.removeall:
```
Comment (pamelafox) [inline scripts/prepdocs.py:387]: Did we lose removeall entirely? I expected it to move outside the function.

Comment (PfisterAn) [reply scripts/prepdocs.py:387]: Its in the __main__ body, it was already there before, thats why you dont see it in the Review.

https://github.com/Azure-Samples/azure-search-openai-demo/blob/36c14b6248b4a5c82e80c6682535990651b25bc6/scripts/prepdocs.py#L397-L405

File: scripts/prepdocs.py (hunk 4 above)
Comment (PfisterAn) [reply scripts/prepdocs.py]: done

File: scripts/prepdocs.py (hunk 5 above)
Comment (PfisterAn) [reply scripts/prepdocs.py]: done

**Comment by PfisterAn** [comment]
Missed to pass the filename towards the split_text function. split_text was setup to utilise the global variable filename, which doesnt exist anymore after introducing the read_files function.

**Review by pamelafox — APPROVED** [review]

**Pull request merged**